windows-agent/internal/daemon/networking.go

```diff
@@ -0,0 +1,101 @@
+package daemon
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+
+	log "github.com/canonical/ubuntu-pro-for-wsl/common/grpc/logstreamer"
+)
+
+// getWslIP returns the loopback address if the networking mode is mirrored or iterates over the network adapters to find the IP address of the WSL one.
+func getWslIP(ctx context.Context, i hostIpconfig, w wslSystemDistro) (net.IP, error) {
+	mode, err := networkingMode(ctx, w)
+	if err != nil {
+		log.Warningf(ctx, "could not determine if WSL network is mirrored (assuming NAT): %v", err)
+		mode = "nat"
+	}
+
+	switch mode {
+	case "mirrored":
+		return net.IPv4(127, 0, 0, 1), nil
+	case "nat":
+		return findWslAdapterIP(i)
+	default:
+		return nil, fmt.Errorf("unknown networking mode: %s", mode)
+	}
+}
+
+func findWslAdapterIP(i hostIpconfig) (net.IP, error) {
+	const targetDesc = "Hyper-V Virtual Ethernet Adapter"
+	const vEthernetName = "vEthernet (WSL"
+
+	head, err := i.getAdaptersAddresses()
+	if err != nil {
+		return nil, err
+	}
+
+	// Filter the adapters by description.
+	var candidates []ipAdapterAddresses
+	for node := head; node != nil; node = node.Next() {
+		desc := node.Description()
+		// The adapter description could be "Hyper-V Virtual Ethernet Adapter #No"
+		if !strings.Contains(desc, targetDesc) {
+			continue
+		}
+
+		candidates = append(candidates, node)
+	}
+
+	if len(candidates) == 1 {
+		return candidates[0].IP(), nil
+	}
+
+	// Desambiguates the adapters by friendly name.
+	for _, node := range candidates {
+		if !strings.Contains(node.FriendlyName(), vEthernetName) {
+			continue
+		}
+
+		return node.IP(), nil
+	}
+
+	return nil, fmt.Errorf("could not find WSL adapter")
+}
+
+// networkingMode detects whether the WSL network is mirrored or not.
+func networkingMode(ctx context.Context, wsl wslSystemDistro) (string, error) {
+	// It does so by launching the system distribution.
+	cmd := wsl.Command(ctx, "wslinfo", "--networking-mode", "-n")
+
+	var stdout, stderr bytes.Buffer
+
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	err := cmd.Run()
+	out := bytes.TrimSpace(stdout.Bytes())
+	if err != nil {
+		return "", fmt.Errorf("%s: error: %v.\n    Stdout: %s\n    Stderr: %s", cmd.Path, err, out, stderr.String())
+	}
+
+	return strings.Split(string(out), "\n")[0], nil
+}
+
+type wslSystemDistro interface {
```

Comment:
I think we probably can drastivally simplify by removing the needs to define interfaces there in getting ` ipConfig` and `wslSystem` in `networking.go`. Those can embbed a mock struct as children, and the implemention in networking.go of `getAdaptersAddresses` for instance use the embedded object to drive the real or mock implementation.

See also my other comments which works in conjunction with this one. WDYT?

Reply:
I didn't quite get this part. Focusing on the `ipConfig` struct, which is dummy, it only serves to allow inject a mock behavior for the Win32 API. You're suggesting to embed the mock inside the production one, but then how are we going to dispatch between the production and the mock versions of `GetAdaptersAddresses`? And embeding on that direction won't let me pass the mock as argument of type `ipConfig`. If I do the other way around, embedding the production one into the mock, then I'll be able to pass the mock where an `ipConfig` object is expected, but there is no means of dispatchig the call to the replacement method of `GetAdaptersAddresses`.

Reply:
I think it will work better if I pass the function as part of the `oprions` struct. The production implementation is a function forwarding to the Win32 API, while the mocked version can be a method bound to a specific instance of the mock type. This way I can avoid the `ipConfig` interface and related types (except for the mock one) and still retain the ability to customize the mock object state (and thus the behavior of the `getAdaptersAddresses` function.

Reply:
sounds good to me. Either that, or we have one big "system" instance where we attach all the side effects and mocks :)